implementations/code_df/is_source_code.py

```diff
@@ -0,0 +1,163 @@
+class IsSourceCode():
+    """
+    A IsSourceCode object is used to help define source code.
+    :param source_code_exclude_list: a list of either
+        file extensions or directories to exclude
+        For example,
+        source_code_exclude_list = ["py", "other", "gitignore"]
+        or source_code_exclude_list = ["tests/", "bin/"]
+    :param algorithm_class: The algorithm class to be used
+    """
+    def __init__(self, source_code_exclude_list=None, algorithm_class=None):
+        self.source_code_exclude_list = source_code_exclude_list
+        self.algorithm_class = algorithm_class
+
+    def check(self, file):
+        """
+        Returns the check method of the algorithm to be used.
+        :param file: a dictionary, an element of commit['files'] list
+            where commit is a structure returned by commit._flatten_data
+
+        :returns Boolean value: True if the file is part of source code,
+            False otherwise
+        """
+        return self.algorithm_class.check(self.source_code_exclude_list, file)
+
+
+class Algorithm:
+    """
+    Parent class to all algorithms.
+    """
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def check(source_code_exclude_list, file):
+        raise NotImplementedError
+
+
+class Naive(Algorithm):
+    """
+    Instantiates an object of Naive. It is one of several
+    "algorithms" used to define source code.
+    """
+    def __init__(self):
+        super().__init__()
```

Comment:
@jgbarah, you mentioned raising a NotImplementedError here. Is that recommended? Would it be better to just not add __init__?

Reply:
I think this is a corner case, since it is very rare to have a class which is not intended to be instantiated (see my other comments about this). In any case, following the "explicit better than implicit", if the class is not intended to be instantiated, I think it is better that `__init__` raises an exception. See [documentation of NotImplementedError](https://docs.python.org/3/library/exceptions.html#NotImplementedError): although it is meant for abstract classes, it is the closer case I can think of.

Reply:
OK, will do!